crates/dbsp/src/operator/communication/exchange.rs

```diff
@@ -31,7 +60,188 @@ use std::{
 // using the id as a key.  If there already is an `Exchange` with this id in
 // the store, created by another worker, a reference to that `Exchange` will
 // be used instead.
-circuit_cache_key!(local ExchangeId<T>(usize => Arc<Exchange<T>>));
+circuit_cache_key!(local ExchangeCacheId<T>(ExchangeId => Arc<Exchange<T>>));
+
+#[tarpc::service]
+trait ExchangeService {
+    async fn exchange(data: Vec<u8>, exchange_id: usize, sender: usize, receiver: usize);
+}
+
+type ExchangeId = usize;
+
+// Maps from an `exchange_id` to the `Inner` that implements the exchange.
+type ExchangeDirectory = Arc<RwLock<HashMap<ExchangeId, Arc<Inner>>>>;
+
+#[derive(Clone)]
+struct ExchangeServer(ExchangeDirectory);
+
+#[tarpc::server]
+impl ExchangeService for ExchangeServer {
+    async fn exchange(
+        self,
+        _: context::Context,
+        data: Vec<u8>,
+        exchange_id: ExchangeId,
+        sender: usize,
+        receiver: usize,
+    ) {
+        let inner = self.0.read().unwrap().get(&exchange_id).unwrap().clone();
+        inner.received(data, sender, receiver).await;
+    }
+}
+
+struct Clients(Vec<(Range<usize>, ExchangeServiceClient)>);
+
+impl Clients {
+    fn new(runtime: &Runtime, local_client: &ExchangeServiceClient) -> Clients {
+        Self(match runtime.layout() {
+            Layout::Solo { n_workers } => vec![(0..*n_workers, local_client.clone())],
+            Layout::Multihost {
+                hosts,
+                local_host_idx,
+            } => hosts
+                .iter()
+                .enumerate()
+                .map(|(idx, host)| {
+                    let client = if idx != *local_host_idx {
+                        let mut transport = connect(host.address, Bincode::default);
+                        transport.config_mut().max_frame_length(usize::MAX);
+                        // XXX Blocking here could cause deadlock?  We should connect
+                        // asynchronously.
+                        sleep(std::time::Duration::from_millis(1000));
```

Comment:
why do you need to sleep?

Reply:
The current code does not retry connection failure, so it's necessary to make it possible to start all the listeners before any of the outgoing connections. A sleep makes that easier to accomplish. This will have to be fixed before this PR is ready to merge, and it's one of the leading reasons why this is a draft PR.

Reply:
The current version does not sleep and does retry connection failure if necessary.